packages/bus-mongodb/src/mongodb-persistence.ts

```diff
@@ -0,0 +1,301 @@
+    includeCompleted = false
+  ): Promise<WorkflowStateType[]> {
+    this.logger.debug("Getting workflow state", {
+      workflowStateName: workflowStateConstructor.name,
+    });
+    const workflowStateName = new workflowStateConstructor().$name;
+    const tableName = resolveQualifiedTableName(workflowStateName);
+    const matcherValue = messageMap.lookup(message, attributes);
+    const workflowStateField = `${WORKFLOW_DATA_FIELD_NAME}.${normalizeProperty(
+      messageMap.mapsTo
+    )}`;
+    const collection = this.database?.collection(tableName);
+    const findObject = {
+      [workflowStateField]: matcherValue,
+    } as any;
+    if (!includeCompleted) {
+      findObject[
+        `${WORKFLOW_DATA_FIELD_NAME}.${normalizeProperty("$status")}`
+      ] = "running";
+    }
+    const documents = await collection?.find(findObject).toArray();
+    this.logger.debug("Querying workflow state", { findObject });
+
+    this.logger.debug("Got workflow state", {
+      resultsCount: documents?.length,
+    });
+
+    const rows = documents?.map((x) => x[WORKFLOW_DATA_FIELD_NAME]) as any[];
+    return rows
+      .map((row) => mapKeys(row, (value, key) => denormalizeProperty(key)))
+      .filter((workflowState) => workflowState !== undefined)
+      .map((workflowState) =>
+        this.coreDependencies.serializer.toClass(
+          workflowState!,
+          workflowStateConstructor
+        )
+      );
+  }
+
+  async saveWorkflowState<WorkflowStateType extends WorkflowState>(
+    workflowState: WorkflowStateType
+  ): Promise<void> {
+    this.logger.debug("Saving workflow state", {
+      workflowStateName: workflowState.$name,
+      id: workflowState.$workflowId,
+    });
+    const collectionName = resolveQualifiedTableName(workflowState.$name);
+
+    const oldVersion = workflowState.$version;
+    const newVersion = oldVersion + 1;
+    const modifiedState = mapKeys(workflowState, (value, key) => {
+      return normalizeProperty(key);
+    });
+    const plainWorkflowState = {
+      ...this.coreDependencies.serializer.toPlain(modifiedState),
+      __version: newVersion,
+    };
+
+    await this.upsertWorkflowState(
+      collectionName,
+      workflowState.$workflowId,
+      plainWorkflowState,
+      oldVersion,
+      newVersion
+    );
+  }
+
+  private async ensureCollectionExists(collectionName: string): Promise<void> {
+    this.logger.debug("Ensuring mongodb collection for workflow state exists", {
+      collectionName,
+    });
+    const collectionExists = await this.database
+      ?.listCollections({ name: collectionName }, { nameOnly: true })
+      .hasNext();
+    if (!collectionExists) {
+      await this.database?.createCollection(collectionName);
+    }
+  }
+
+  private async ensureIndexesExist(
+    collectionName: string,
+    messageWorkflowMappings: MessageWorkflowMapping<Message, WorkflowState>[]
+  ): Promise<void> {
+    const collection = this.database?.collection(collectionName);
```

Comment:
this actually runs before `initialize()`, so `this.database` is always undefined and none of these indexes are actually created. I'd recommend:

1. don't use optional chaining (`this.database?.`) but instead the non-null assertion throughout this method eg:
```suggestion
    const collection = this.database!.collection(collectionName);
```

Reply:
Can you tell me in which case the initialize method runs after initializeWorkflow()? @adenhertog

Reply:
`initializeWorkflow()` always runs before `initialize()`. You should be able to see this at `workflow-registry.ts:112` and `workflow-registry.ts:120`

Reply:
Fixed plz check the solution @adenhertog